Add verifications of `ClientRequestContext`s to `RetryClientTest` and `RetryingRpcClientTest`

schiemon opened the pull request:
## Motivation:

`RetryingClientTest` and `RetryingRpcClientTest` validate the respective retrying clients under various (failure) conditions.  
However, not all test cases check whether the client-side request log is properly propagated by the client, e.g., whether:
- the log is completed,
- all child logs are added and completed as well,
- all child logs have the expected status code and exception.

Verifying the logs becomes even more important for #6252, where the retrying client must ensure it properly completes concurrent pending requests—either when a response is received or when a request fails unrecoverably, thereby also failing all other pending requests. Aforementioned tests would increase our confidence that the upcoming hedging client does not cause regressions.

## Modifications:

All changes are test-only.

- Introduced a small testing utility (`RequestContextUtils`) for verifying `ClientRequestContext`s, especially the log, in a concise way.
- Extended test cases in `RetryingClientTest` and `RetryingRpcClientTest` by incorporating verifications using that utility class.

## Result:

- 26 + 10 test cases now also verify `ClientRequestContext` under various failure scenarios.
- `RequestContextUtils` now helps in writing concise tests that validate `ClientRequestContext`, e.g., for request hedging.

schiemon committed:
test: verify valid client request context after retrying test

schiemon committed:
test: add more detailed verification of child logs

schiemon committed:
test: add client request context verifications to `RetryingRpcClientTest`

schiemon committed:
refactor: remove unused logQueue from `RetryingRpcClientTest`

schiemon committed:
refactor: remove unused verifyRequestCause(Throwable) from `RequestContextUtils`

jrhee17 commented on core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientTest.java:
```diff
@@ -810,10 +1078,23 @@ void exceptionInRuleWithContent() {
-        assertThatThrownBy(client.get("/").aggregate()::join)
-                .isInstanceOf(CompletionException.class)
-                .hasCauseReference(exception);
+        final WebClient client = client(rule,
+                                        10000, 0, 100);
```
nit; 
```suggestion
        final WebClient client = client(rule, 10000, 0, 100);
```

jrhee17 reviewed (commented):
I appreciate the additional validation in retrying tests. I've reviewed mainly `RequestContextUtils` mainly considering how we can use this utility more extensively within the code-base.

schiemon committed:
refactor: rename `verifyChildLog` to `verifyLog`

schiemon committed:
refactor: rename `verifyExactlyOneValid` to `verifyExactlyOneVerifierValid`

schiemon committed:
refactor: remove skipping of verifiers when there are no child logs

schiemon committed:
style: remove unnecessary line break

schiemon committed:
refactor: rename `verifyAllValid` to `verifyAllVerifierValid`

schiemon committed:
refactor: enforce that users provide a parent log verifier

schiemon committed:
refactor: improve internal naming

schiemon commented:
@jrhee17 Thank you for the review 🤙 . I addressed your comments. I am not sure if you like me to close the comments myself or leave them open. Feel free to educate me

schiemon committed:
Merge branch 'main' into verify-client-request-context-in-retrying-client-test
